eng/common/scripts/Generate-DocIndex.ps1

```diff
@@ -0,0 +1,147 @@
+# Generates an index page for cataloging different versions of the Docs
+[CmdletBinding()]
+Param (
+    $RepoRoot,
+    $DocGenDir,
+    $lang = "c", # Take c for example. should be empty string
+    $packageNameRegex = "",
+    $packageNameReplacement = "",
+    $blobPath = "docs" # This is a setting for c. should be empty string for all other repo.
+)
+. (Join-Path $PSScriptRoot ../common/scripts/Package-Properties.ps1)
+
+# A mapping is used to fill in 1. Title of home page, 2. The azure storage link
+$LangMapping = @{
+    ".net"       = @("NET", "dotnet")
+    "java"       = @("Java", "java")
+    "javascript" = @("JavaScript", "javascript")
+    "python"     = @("Python", "python")
+    "c"          = @("C", "c")
+    "cpp"        = @("CPP", "cpp")
+}
+
+# The sequence of Bom bytes differs by different encoding. 
+# The helper function here is only to strip the utf-8 encoding system as it is used by blob storage list api.
+# Return the original string if not in BOM utf-8 sequence.
+function removeBomFromString($bomAwareString) {
+    if ($bomAwareString.length -le 3) {
+        return $bomAwareString
+    }
+    $bomPatternByteArray = [byte[]] (0xef, 0xbb, 0xbf)
+    # The default encoding for powershell is ISO-8859-1, so converting bytes with the encoding.
+    $bomAwareBytes = [Text.Encoding]::GetEncoding(28591).GetBytes($bomAwareString.Substring(0, 3))
+    if (@(Compare-Object $bomPatternByteArray $bomAwareBytes -SyncWindow 0).Length -eq 0) {
+        return $bomAwareString.Substring(3)
+    }
+    return $bomAwareString
+}
+Write-Verbose "Name Reccuring paths with variable names"
+$DocFxTool = "${RepoRoot}/docfx/docfx.exe"
+$DocOutDir = "${RepoRoot}/docfx_project"
+
+Write-Verbose "Initializing Default DocFx Site..."
+& "${DocFxTool}" init -q -o "${DocOutDir}"
+
+Write-Verbose "Copying template and configuration..."
+New-Item -Path "${DocOutDir}" -Name "templates" -ItemType "directory" -Force
+Copy-Item "${DocGenDir}/templates/*" -Destination "${DocOutDir}/templates" -Force -Recurse
+Copy-Item "${DocGenDir}/docfx.json" -Destination "${DocOutDir}/" -Force
+$YmlPath = "${DocOutDir}/api"
+New-Item -Path $YmlPath -Name "toc.yml" -Force
+
+Write-Verbose "Reading artifact from storage blob ..."
+$metadata = GetMetaData -lang $lang
```

Comment:
probably worth doing the groupid metadata filtering just once here instead of every time through the loop.